src/transformation/transformation.ts

```diff
@@ -3,9 +3,305 @@ import {
+ * transform and assign children.
+ * If no factory is provided for a source node type, a GenericNode is created, and the processing of the subtree stops
+ * there.
+ */
+export class ASTTransformer {
+    private readonly _issues: Issue[];
+    private readonly allowGenericNode: boolean;
+
+    get issues() : Issue[] {
+        return this._issues;
+    }
+
+    /**
+     * Factories that map from source tree node to target tree node.
+     */
+    private factories = new Map<any, NodeFactory<any, any>>();
+    private knownClasses = new Map<string, any>(); //<-- TODO: not sure about this piece of the port from Kolasu
+
+    /**
+     * @param issues Additional issues found during the transformation process.
+     * @param allowGenericNode Use GenericNode as a strategy for missing factories for nodes.
+     */
+    constructor(issues: Issue[] = [], allowGenericNode = true) {
+        this._issues = issues;
+        this.allowGenericNode = allowGenericNode;
+    }
+
+    addIssue(message: string, severity: IssueSeverity = IssueSeverity.ERROR, position?: Position) : Issue {
+        const issue = Issue.semantic(message, severity, position);
+        this._issues.push(issue);
+        return issue;
+    }
+
+    transform(source?: any, parent?: Node) : Node | undefined {
+        if (source == undefined)
+            return undefined;
+
+        if (Array.isArray(source))
+            throw Error("Mapping error: received collection when value was expected");
+
+        const factory: NodeFactory<any, any> | undefined = this.getNodeFactory(source);
+        let node: Node | undefined;
+
+        if (factory != undefined) {
+            node = this.makeNode(factory, source);
+
+            if (node == undefined)
+                return undefined;
+
+            Object.keys(node).forEach(propertyName => {
+                const nodeClass = Object.getPrototypeOf(node).constructor;
+                const prefix = isClassType(nodeClass) ? `${nodeClass.name}#` : "";
+                const childKey: string = prefix + propertyName;
+                const childNodeFactory = factory.children.get(childKey);
+                if (childNodeFactory) {
+                    if (childNodeFactory !== NO_CHILD_NODE) {
+                        this.setChild(childNodeFactory, source, node!, propertyName);
+                    }
+                } else {
+                    // TODO: case where MAPPED is used
+                    factory.children.set(childKey, NO_CHILD_NODE);
+                }
+            });
+
+            factory.finalizer(node);
+            node.parent = parent;
+        }
+        else {
+            if (this.allowGenericNode) {
+                const origin : Origin | undefined = this.asOrigin(source);
+                node = new GenericNode(parent).withOrigin(origin);
+                this._issues.push(
+                    Issue.semantic(
+                        `Source node not mapped: ${Object.getPrototypeOf(source).constructor.name}`,
```

Comment:
I would rather use getNodeDefinition here, as the name will be mangled when using a minifier.